File: app/src/main/java/org/schabi/newpipe/local/history/HistoryEntryAdapter.java
package org.schabi.newpipe.local.history;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;

import org.schabi.newpipe.util.Localization;

import java.text.DateFormat;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Date;


/**
 * Adapter for history entries
 * @param <E> the type of the entries
 * @param <VH> the type of the view holder
 */
public abstract class HistoryEntryAdapter<E, VH extends RecyclerView.ViewHolder> extends RecyclerView.Adapter<VH> {

    private final ArrayList<E> mEntries;
    private final DateFormat mDateFormat;
    private final Context mContext;
    private OnHistoryItemClickListener<E> onHistoryItemClickListener = null;


    public HistoryEntryAdapter(Context context) {
        super();
        mContext = context;
        mEntries = new ArrayList<>();
        mDateFormat = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.MEDIUM,
                Localization.getPreferredLocale(context));
    }

    public void setEntries(@NonNull Collection<E> historyEntries) {
        mEntries.clear();
        mEntries.addAll(historyEntries);
        notifyDataSetChanged();
    }

    public Collection<E> getItems() {
        return mEntries;
    }

    public void clear() {
        mEntries.clear();
        notifyDataSetChanged();
    }

    protected String getFormattedDate(Date date) {
        return mDateFormat.format(date);
    }

    protected String getFormattedViewString(final long viewCount) {
        return Localization.shortViewCount(mContext, viewCount);
    }

    @Override
    public int getItemCount() {
        return mEntries.size();
    }

    @Override
    public void onBindViewHolder(VH holder, int position) {
        final E entry = mEntries.get(position);
        holder.itemView.setOnClickListener(v -> {
            if(onHistoryItemClickListener != null) {
                onHistoryItemClickListener.onHistoryItemClick(entry);
            }
        });

        holder.itemView.setOnLongClickListener(view -> {
            if (onHistoryItemClickListener != null) {
                onHistoryItemClickListener.onHistoryItemLongClick(entry);
                return true;
            }
            return false;
        });

        onBindViewHolder(holder, entry, position);
    }

    @Override
    public void onViewRecycled(VH holder) {
        super.onViewRecycled(holder);
        holder.itemView.setOnClickListener(null);
    }

    abstract void onBindViewHolder(VH holder, E entry, int position);

    public void setOnHistoryItemClickListener(@Nullable OnHistoryItemClickListener<E> onHistoryItemClickListener) {
        this.onHistoryItemClickListener = onHistoryItemClickListener;
    }

    public boolean isEmpty() {
        return mEntries.isEmpty();
    }

    public interface OnHistoryItemClickListener<E> {
        void onHistoryItemClick(E item);
        void onHistoryItemLongClick(E item);
    }
}

File: app/src/main/java/org/schabi/newpipe/local/history/HistoryEntryAdapter.kt
package org.schabi.newpipe.local.history

import android.content.Context
import android.view.View
import android.view.View.OnLongClickListener
import androidx.recyclerview.widget.RecyclerView
import org.schabi.newpipe.util.Localization
import java.text.DateFormat
import java.util.Date

/**
 * This is an adapter for history entries.
 *
 * @param <E>  the type of the entries
 * @param <VH> the type of the view holder
</VH></E> */
abstract class HistoryEntryAdapter<E, VH : RecyclerView.ViewHolder?>(private val mContext: Context) : RecyclerView.Adapter<VH>() {
    private val mEntries: ArrayList<E>
    private val mDateFormat: DateFormat
    private var onHistoryItemClickListener: OnHistoryItemClickListener<E>? = null

    init {
        mEntries = ArrayList()
        mDateFormat = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.MEDIUM,
                Localization.getPreferredLocale(mContext))
    }

    fun setEntries(historyEntries: Collection<E>) {
        mEntries.clear()
        mEntries.addAll(historyEntries)
        notifyDataSetChanged()
    }

    val items: Collection<E>
        get() {
            return mEntries
        }

    fun clear() {
        mEntries.clear()
        notifyDataSetChanged()
    }

    protected fun getFormattedDate(date: Date?): String {
        return mDateFormat.format(date)
    }

    protected fun getFormattedViewString(viewCount: Long): String? {
        return Localization.shortViewCount(mContext, viewCount)
    }

    public override fun getItemCount(): Int {
        return mEntries.size
    }

    public override fun onBindViewHolder(holder: VH, position: Int) {
        val entry: E = mEntries.get(position)
        holder!!.itemView.setOnClickListener(View.OnClickListener({ v: View? ->
            if (onHistoryItemClickListener != null) {
                onHistoryItemClickListener!!.onHistoryItemClick(entry)
            }
        }))
        holder.itemView.setOnLongClickListener(OnLongClickListener({ view: View? ->
            if (onHistoryItemClickListener != null) {
                onHistoryItemClickListener!!.onHistoryItemLongClick(entry)
                return@setOnLongClickListener true
            }
            false
        }))
        onBindViewHolder(holder, entry, position)
    }

    public override fun onViewRecycled(holder: VH) {
        super.onViewRecycled(holder)
        holder!!.itemView.setOnClickListener(null)
    }

    abstract fun onBindViewHolder(holder: VH, entry: E, position: Int)
    fun setOnHistoryItemClickListener(
            onHistoryItemClickListener: OnHistoryItemClickListener<E>?) {
        this.onHistoryItemClickListener = onHistoryItemClickListener
    }

    val isEmpty: Boolean
        get() {
            return mEntries.isEmpty()
        }

    open interface OnHistoryItemClickListener<E> {
        fun onHistoryItemClick(item: E)
        fun onHistoryItemLongClick(item: E)
    }
}
